File: tests/fixtures/python/simple_function.py
"""
Standalone typed functions for Python with modern 3.10+ union syntax.
"""

from typing import Sequence


def add_numbers(a: int | float, b: int | float) -> int | float:
    """Add two numbers supporting ints and floats."""
    return a + b


def format_user_name(first_name: str, last_name: str, prefix: str | None = None) -> str:
    """Format a user's full name with an optional honorific prefix."""
    full = f"{first_name.strip()} {last_name.strip()}".strip()
    if prefix and prefix.strip():
        return f"{prefix.strip()} {full}"
    return full


def parse_identifier(raw: str | int | bytes) -> str:
    """Normalize various identifier input formats into a canonical string."""
    if isinstance(raw, bytes):
        return raw.decode("utf-8", errors="replace").strip()
    if isinstance(raw, int):
        return f"id_{raw:08d}"
    return str(raw).strip()


def calculate_discount(price: float, percentage: float) -> float:
    """Calculate discounted price with input validation."""
    if price < 0.0:
        raise ValueError("Price cannot be negative")
    if not (0.0 <= percentage <= 100.0):
        raise ValueError("Percentage must be between 0 and 100")
    discount = (price * percentage) / 100.0
    return round(price - discount, 2)


def find_first_matching(items: Sequence[str | int], query: str | int) -> int | None:
    """Find the index of the first matching item in a sequence."""
    for idx, item in enumerate(items):
        if item == query:
            return idx
    return None

File: tests/fixtures/typescript/simple_function.ts
/**
 * Standalone typed helper functions for testing AST extraction of pure functions.
 */

export function addNumbers(a: number, b: number): number {
    return a + b;
}

export function formatUserName(firstName: string, lastName: string, prefix?: string): string {
    const fullName = `${firstName.trim()} ${lastName.trim()}`.trim();
    if (prefix && prefix.trim().length > 0) {
        return `${prefix.trim()} ${fullName}`;
    }
    return fullName;
}

export function calculateDiscount(price: number, percentage: number): number {
    if (price < 0) {
        throw new RangeError("Price cannot be negative");
    }
    if (percentage < 0 || percentage > 100) {
        throw new RangeError("Percentage must be between 0 and 100");
    }
    const discount = (price * percentage) / 100;
    return Math.round((price - discount) * 100) / 100;
}

export function clamp(value: number, min: number, max: number): number {
    if (min > max) {
        throw new Error("Min cannot be greater than Max");
    }
    return Math.min(Math.max(value, min), max);
}

export function isNonEmptyString(input: unknown): input is string {
    return typeof input === "string" && input.trim().length > 0;
}

export function generateSlug(text: string): string {
    return text
        .toLowerCase()
        .trim()
        .replace(/[^\w\s-]/g, "")
        .replace(/[\s_-]+/g, "-")
        .replace(/^-+|-+$/g, "");
}
